dm-agent/src/main.rs

```diff
@@ -0,0 +1,354 @@
+    )]
+    InvalidMeasurementPayloadFormat(String, String),
+
+    #[error("Invalid measurement value: {0} received on topic: {1}. Must be a number")]
+    InvalidMeasurementValue(String, String),
+
+    #[error("Invalid Thin Edge JSON measurement format")]
+    InvalidThinEdgeJsonError(#[from] thin_edge_json::group::MeasurementGrouperError),
+
+    #[error(transparent)]
+    ThinEdgeJsonSerializationError(#[from] ThinEdgeJsonSerializationError),
+}
+
+struct CollectdMessage {
+    metric_group_key: String,
+    metric_key: String,
+    metric_value: f64,
+}
+
+impl TryFrom<Message> for CollectdMessage {
+    type Error = Error;
+
+    fn try_from(mqtt_message: Message) -> Result<Self, Error> {
+        let topic = mqtt_message.topic.name.clone();
+        let split = topic.split('/').collect::<Vec<&str>>();
+        if split.len() == 4 {
+            let metric_group_key = split[METRIC_GROUP_KEY_POS].into();
+            let metric_key = split[METRIC_KEY_POS].into();
+
+            let payload = mqtt_message.payload;
+            let payload = from_utf8(&payload)
+                .map_err(|_err| Error::InvalidMeasurementPayload(topic.clone()))?; //TODO: avoid topic clone?
+            let split = payload.split(':').collect::<Vec<&str>>();
+
+            if split.len() == 2 {
+                let metric_value = split[1]
+                    .trim_end_matches(char::from(0))
+                    .parse::<f64>()
+                    .map_err(|_err| {
+                        Error::InvalidMeasurementValue(split[1].into(), topic.clone())
+                    })?;
+
+                Ok(Self {
+                    metric_group_key,
+                    metric_key,
+                    metric_value,
+                })
+            } else {
+                return Err(Error::InvalidMeasurementPayloadFormat(
+                    payload.into(),
+                    topic.clone(),
+                ));
+            }
+        } else {
+            //TODO: Why explicit return
+            return Err(Error::InvalidMeasurementTopic(topic));
+        }
+    }
+}
+
+#[derive(Debug)]
+struct ThinEdgeMessage {
+    message_grouper: MeasurementGrouper,
+}
+
+impl ThinEdgeMessage {
+    pub fn new() -> Self {
+        Self {
+            message_grouper: MeasurementGrouper::new(),
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.message_grouper.is_empty()
+    }
+
+    pub fn get_measurement_value(
+        &self,
+        group_key: Option<&str>,
+        measurement_key: &str,
+    ) -> Option<f64> {
+        self.message_grouper
+            .get_measurement_value(group_key, measurement_key)
+    }
+
+    pub fn add_measurement(&mut self, collectd_message: CollectdMessage) -> Result<(), Error> {
+        if self.message_grouper.is_empty() {
+            let _ = self
+                .message_grouper
+                .timestamp(&current_timestamp())
+                .map_err(|err| error!("{}", err));
```

Comment:
The fact that the timestamp is assigned internally will make unclear the relationship between this timestamp and the collection time window.

Reply:
Not sure if I understood that suggestion completely. The timestamp is generated on the receipt of the very first message and NOT when the collection window starts. It was done this way due to the limitation around the collection window trigger in the current implementation which can lead to the window start time and first message receipt time to be different. Is that what you were highlighting? In that case, once we fix the collection trigger issue, this won't be an issue anymore.